gsoc2020/20200303-health-check-wjayesh.md

```diff
@@ -0,0 +1,254 @@
+ 
+Minimal assumptions about the users’ Kubernetes deployment are made and the primary **focus** is to build a program that could easily be adopted by more teams (which may have different preferences of deployment). 
+ 
+In any case, the binary will accomplish the following fundamental tasks: 
+
+
+- Use `kubectl` to extract information about the cluster and its services and pods.  
+ 
+  This will help determine the procedure to follow and also facilitate the subsequent actions to take. This step can be completed only when the application can connect and communicate with the API server. The Go [client](https://github.com/kubernetes/client-go) for Kubernetes has packages that can discover and authenticate with the API server.  
+ 
+- Find the IP address (cluster IP and pod IP) of the CoreDNS service and pods.  
+ 
+  These IP addresses will be used for the `dig` operation in the next steps. It is critical to use both cluster IP and the pod IP. Although it may seem that pinging cluster IP would be sufficient, in real production systems, any point of the system could fail. We cannot make the assumption that Kubernetes networking (forwarding requests from the Service IP to the Pods) will work all the time.  
+ 
+  Apart from this, it’ll be useful to know which Pods are sick, so they can be remedied individually.  
+ 
+- Use `dig` to ping the cluster IP and pod IPs through DNS. 
+ 
+  `dig` is a tool that can be used to look up DNS servers using their IP addresses and port numbers (default is 53). The output of the command execution will help determine whether the DNS service is healthy or not.  
+ 
+- Restart CoreDNS pods through the API server if pods are up and running but DNS reply does not match the expectation. 
+
+
+ 
+### Strategies for Different Deployment Scenarios 
+
+I have listed out four possible scenarios that the binary might be deployed in. Out of these, three have a determinate solution and can be implemented right away. The last one will require some hacking and also relies on a feature which is under development. 
+- **Binary Deployed as a Pod** 
+
+
+  Here, the team decides to launch the binary as a health check pod using `kubectl` and a YAML configuration. The steps listed above can then be directly executed with no additional configuration or modifications required.  
+
+
+- **Binary Deployed Externally but Creating/Deleting Pods Inside the Cluster Allowed** 
+![CoreDNSPod](https://user-images.githubusercontent.com/37150991/75658062-9220f900-5c8d-11ea-8c13-f48fbcecc37a.jpeg)
+ 
+ 
+  Here, the binary itself lives as an external program but will dynamically create a lightweight pod in the cluster (through communication with the API server). The only purpose of this pod inside the cluster would be to execute the dig command (and maybe other debug commands) that the external Golang application sends to this pod. 
```

Comment:
@WJayesh , let's try "Pros/Cons" pattern here. That is describe "Pros/Cons" for each of the deployment patterns.
You could also create a table for each of the deployment options, e.g.:

```
| **Pros** | **Cons** |
| --- | --- |
| lightweight pod | TBD |
```